x/rollapp/keeper/liveness.go

```diff
@@ -0,0 +1,154 @@
+package keeper
+
+import (
+	"time"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+	"github.com/osmosis-labs/osmosis/v15/osmoutils"
+)
+
+/*
+This file has the logic for slashing rollapps based on liveness requirements (time since last update).
+It will trigger slash/jail operations through the x/sequencers module, at intervals decided by parameters.
+See ADR for more info https://www.notion.so/dymension/ADR-x-Sequencer-Liveness-Slash-Phase-1-5131b4d557e34f4498855831f439d218
+*/
+
+// NextSlashOrJailHeight returns the next height on the HUB to slash or jail the rollapp
+// It will respect all parameters passed in.
+func NextSlashOrJailHeight(
+	hubBlockInterval time.Duration, // average time between hub blocks
+	slashTimeNoUpdate time.Duration, // time until first slash if not updating
+	slashInterval time.Duration, // gap between slash if still not updating
+	jailTime time.Duration, // time until jail if not updating
+	heightHub int64, // current hub height
+	heightLastRollappUpdate int64, // when was the rollapp last updated
+) (
+	heightEvent int64, // hub height to schedule event
+	isJail bool, // is it a jail event? (false -> slash)
+) {
+	// how long has the rollapp been down already?
+	downTime := time.Duration(heightHub-heightLastRollappUpdate) * hubBlockInterval
+	// when should we schedule the next slash/jail, in terms of down time duration?
+	targetDuration := slashTimeNoUpdate
+	if slashTimeNoUpdate < downTime {
+		targetDuration += ((downTime-slashTimeNoUpdate)/slashInterval + 1) * slashInterval
+	}
+	heightEvent = heightLastRollappUpdate + int64((targetDuration+hubBlockInterval-1)/hubBlockInterval)
+	isJail = jailTime <= targetDuration
+	return
+}
+
+// CheckLiveness will slash or jail any sequencers for whom their rollapp has been down
+// and a slash or jail event is due. Run in end block.
+func (k Keeper) CheckLiveness(ctx sdk.Context) {
+	h := ctx.BlockHeight()
+	events := k.GetLivenessEvents(ctx, &h)
+	for _, e := range events {
+		err := osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+			return k.HandleLivenessEvent(ctx, e)
+		})
+		if err != nil {
+			k.Logger(ctx).Error(
+				"Check liveness event",
+				"event", e,
+				"err", err,
+			)
+		}
+	}
+}
+
+// HandleLivenessEvent will slash or jail and then schedule a new event in the future.
+func (k Keeper) HandleLivenessEvent(ctx sdk.Context, e types.LivenessEvent) error {
+	if e.IsJail {
+		err := k.sequencerKeeper.JailLiveness(ctx, e.RollappId)
+		if err != nil {
+			return errorsmod.Wrap(err, "jail liveness")
+		}
+	} else {
+		err := k.sequencerKeeper.SlashLiveness(ctx, e.RollappId)
+		if err != nil {
+			return errorsmod.Wrap(err, "slash liveness")
+		}
+	}
+
+	ra := k.MustGetRollapp(ctx, e.RollappId)
+	k.DelLivenessEvents(ctx, ra.LivenessEventHeight, ra.RollappId)
+	k.ScheduleLivenessEvent(ctx, &ra)
+	k.SetRollapp(ctx, ra)
```

Comment:
this could go in its own method: ~`RescheduleLivenessEvent`

Reply:
done